pkg/controller/awsmachine/awsmachine_controller.go

```diff
@@ -153,116 +155,92 @@ func (r *ReconcileAWSMachine) Reconcile(request reconcile.Request) (reconcile.Re
 }
 
 func (r *ReconcileAWSMachine) reconcile(ctx context.Context, scope *scope.MachineScope) error {
-	exist, err := r.exists(scope)
-	if err != nil {
-		klog.Errorf("Failed to check if Machine %q infrastructure exists: %v", scope.Name(), err)
-		return err
-	}
-
-	// Reconcile ProviderID.
-	if pid := scope.GetProviderID(); pid == nil || *pid == "" {
-		scope.SetProviderID(fmt.Sprintf("aws:////%s", *scope.GetInstanceID()))
-	}
-
-	if exist {
-		scope.Info("Reconciling Machine triggers idempotent update")
-		return r.update(scope)
-	}
-
-	// Machine resource created. Machine does not yet exist.
-	scope.Info("Reconciling Machine triggers idempotent create")
-	return r.create(scope)
-}
-
-// create creates a machine and is invoked by the machine controller.
-func (r *ReconcileAWSMachine) create(scope *scope.MachineScope) error {
 	if scope.Parent.Cluster.Annotations[infrav1.AnnotationClusterInfrastructureReady] != infrav1.ValueReady {
 		scope.Info("Cluster infrastructure is not ready yet - requeuing machine")
 		return &capierrors.RequeueAfterError{RequeueAfter: waitForClusterInfrastructureReadyDuration}
 	}
 
 	ec2svc := ec2.NewService(scope.Parent)
 
-	scope.Info("Retrieving machines for cluster")
-	machineList := &clusterv1.MachineList{}
-	if err := r.List(context.Background(), machineList, scope.Parent.ListOptionsLabelSelector()); err != nil {
-		return errors.Wrapf(err, "failed to retrieve machines in cluster %q", scope.Parent.Name())
+	// Get or create the instance.
+	instance, err := r.getOrCreate(scope, ec2svc)
+	if err != nil {
+		return err
 	}
 
-	controlPlaneMachines := util.GetControlPlaneMachinesFromList(machineList)
-	if len(controlPlaneMachines) == 0 {
-		scope.Info("No control plane machines exist yet - requeuing")
-		return &capierrors.RequeueAfterError{RequeueAfter: waitForControlPlaneMachineExistenceDuration}
+	// Set an error message if we couldn't find the Machine.
+	if instance == nil {
+		scope.SetErrorMessage(errors.New("EC2 instance cannot be found"))
+		return nil
 	}
 
-	// Create the Machine.
-	i, err := ec2svc.CreateOrGetMachine(scope)
-	if err != nil {
-		return errors.Errorf("failed to create or get machine: %+v", err)
-	}
+	// Make sure Spec.ProviderID is always set.
+	scope.SetProviderID(fmt.Sprintf("aws:////%s", instance.ID))
 
-	scope.SetInstanceID(i.ID)
-	scope.SetInstanceState(i.State)
+	// Proceed to reconcile the AWSMachine state.
+	scope.SetInstanceState(instance.State)
 	scope.SetAnnotation("cluster-api-provider-aws", "true")
```

Comment:
might want to add a TODO here to remove this when it is no longer needed for `clusterctl`